### ISSUE TITLE
Inconsistent search results

### ISSUE THREAD

**Issue by MartyLake** [MusicPlayerDaemon/MPD#820]
### Describe the bug

When using mpc or cantata to search for an artist that exists in my library, I get inconsistent search results. 

## Expected Behavior

Everytime I search for a specific artist, I get all this artist's songs

## Actual Behavior

Sometimes I get 0 result, and I try again 5 seconds later and I get all this artist's songs, then 3 seconds later I try again and 0 result.

## Version

`Music Player Daemon 0.21.21 (0.21.21)` from https://www.musicpd.org/download/win32/0.21.21/mpd.exe

## Log

```
C:\Users\martylake>mpc --password=redacted search any Oliver
A-Trak/Tuna Melt - EP/04 Disco Nap.flac
Oliver/Dirty Talk/01 Dirty Talk.mp3
Oliver/Full Circle/01 Portrait.mp3
Oliver/Full Circle/02 Ottomatic.mp3
Oliver/Full Circle/03 Chemicals.mp3
Oliver/Full Circle/04 Love Like This.mp3
Oliver/Full Circle/05 Last Forever.mp3
Oliver/Full Circle/06 Heart Attack.mp3
Oliver/Full Circle/07 One.mp3
Oliver/Full Circle/08 Space & Sound.mp3
Oliver/Full Circle/09 Falling Back.mp3
Oliver/Full Circle/10 Go With It.mp3
Oliver/Full Circle/11 Wherever We Are.mp3
Oliver/Full Circle/12 Electrify.mp3
Oliver/Full Circle/13 Heterotopia.mp3
Oliver/Full Circle/14 At Night.mp3
Oliver/Full Circle/15 Full Circle.mp3

C:\Users\martylake>mpc --password=redacted search any oliver

C:\Users\martylake>mpc --password=redacted search any Oliver

C:\Users\martylake>mpc --password=redacted search any Oliver
```

corresponding verbose logs :
```
Apr 02 15:16 : client: [9] opened from 127.0.0.1:62836
Apr 02 15:16 : client: [9] process command "password "redacted""
Apr 02 15:16 : client: [9] command returned 0
Apr 02 15:16 : client: [9] process command list
Apr 02 15:16 : client: process command "tagtypes "clear""
Apr 02 15:16 : client: command returned 0
Apr 02 15:16 : client: process command "search any "Oliver""
Apr 02 15:16 : client: command returned 0
Apr 02 15:16 : client: [9] process command list returned 0
Apr 02 15:16 : client: [9] closed
```

**Comment by MaxKellermann** [comment]
Your log of 4 mpc calls contains just 1 client connect. So there appears to be a connection problem. (Or your log is incomplet.)

**Comment by MartyLake** [comment]
Yes, I forgot to mention that IRL it happened like this:
"I tried to search, I got inconsistent results, I updated mpd, I tried to reproduce, it did reproduce, I opened the new issue, and I tried to replicate, and now I don't get any result any more (0 match). So I went back to search the logs and pasted what I though would be meaningful for this issue."

I tried again right now, and I can say that I correctly get one client connect per mpc call.

**Comment by MaxKellermann** [comment]
And which log did you paste? The one with a result or the one with empty result?

**Comment by MartyLake** [comment]
The one with empty result.

Le jeu. 2 avr. 2020 à 16:33, Max Kellermann <notifications@github.com> a
écrit :

**Comment by MaxKellermann** [comment]
So... then, how does a log with a non-empty result look like?
Does that happen only with this query or also with other queries? What about directory listings?
Does this happen with other clients, e.g. ncmpc?
Are you aware that your search query does not search for artists, but matches all tags? Your text says you're looking for artists, but that's not what your query does.

**Comment by MartyLake** [comment]
Thank you very much for helping me on this. I will try to answer all your suggestions, hopefully I can find a workaround. If the current format is not accurate enough, or can be formatted differently, please let me know:

## So... then, how does a log with a non-empty result look like?

mpc:
```
C:\Users\martylake>mpc --password=redacted search any Dance
_/1975/00.mp3
3LAU/Vacayvitamins.com/05 Girls Who Save The World.mp3
3LAU/Vacayvitamins.com/08 Yacht Week In America.mp3
3V/06 HEART OF THE BEAT.mp3
<... redacted plenty of results for this query>
```
logs:
```
Apr 02 17:08 : client: [22] opened from 127.0.0.1:64436
Apr 02 17:08 : client: [22] process command "password "redacted""
Apr 02 17:08 : client: [22] command returned 0
Apr 02 17:08 : client: [22] process command list
Apr 02 17:08 : client: process command "tagtypes "clear""
Apr 02 17:08 : client: command returned 0
Apr 02 17:08 : client: process command "search any "Dance""
Apr 02 17:08 : client: command returned 0
Apr 02 17:08 : client: [22] process command list returned 0
Apr 02 17:08 : client: [22] closed
```


## Does that happen only with this query or also with other queries? What about directory listings?

mpc:
```
C:\Users\martylake>mpc --password=redacted ls Oliver
Oliver/Dirty Talk
Oliver/Full Circle
```
logs:
```
Apr 02 17:11 : client: [23] opened from 127.0.0.1:64456
Apr 02 17:11 : client: [23] process command "password "redacted""
Apr 02 17:11 : client: [23] command returned 0
Apr 02 17:11 : client: [23] process command "lsinfo "Oliver""
Apr 02 17:11 : client: [23] command returned 0
Apr 02 17:11 : client: [23] closed
```

## Does this happen with other clients, e.g. ncmpc?

Using WSL, I installed ncmpc. I have trouble doing a "any" search from ncmpc, but I managed to search for the artist Oliver and got no result. I tried a "genre" "Dance" and got one track. Then I tried it on Cantata, got lots of results for "genre" "Dance", tried a "any" "Oliver", got nothing, tried again "genre" "Dance" and got nothing.

<details><summary>screenshots</summary>
<p>

![mpd empty search any oliver cantata](https://user-images.githubusercontent.com/1246794/78268416-71f08e00-7508-11ea-9639-d57346e9b3e0.PNG)
Tous == any in french

![mpd empty search artist oliver](https://user-images.githubusercontent.com/1246794/78268420-73ba5180-7508-11ea-8eae-4b22429831cb.PNG)
![mpd search genre dance cantata empty](https://user-images.githubusercontent.com/1246794/78268422-73ba5180-7508-11ea-9193-addbf86b904c.PNG)
![mpd search genre dance cantata](https://user-images.githubusercontent.com/1246794/78268427-7452e800-7508-11ea-9236-4c75efe69117.PNG)
![mpd search genre dance ncmpc](https://user-images.githubusercontent.com/1246794/78268439-79b03280-7508-11ea-8cc4-3033b2c637f4.PNG)

</p>
</details>

So yes, it happens with other clients.

## Are you aware that your search query does not search for artists, but matches all tags? Your text says you're looking for artists, but that's not what your query does.

Yes I am aware that my search query matches all tags. Changing the type of query did not change the outcome of the searches, that is inconsistent results.

**Comment by MartyLake** [comment]
Any tips on how to diagnose any further ? I am not sure how to use the search in mpd. Is it something peculiar with my setup ? Are any other users experiencing this ?

**Comment by geneticdrift** [comment]
Just wanted to point out that two of your logs show:
`client: process command "tagtypes "clear""`
just before the search command. So that may explain zero results.

To diagnose further I'd suggest doing manual queries directly without a client using `nc` or `telnet`.
For example:
<pre>
echo "search any \"Dance\"" | nc mpd_hostname 6600
</pre>

**Comment by MartyLake** [comment]
Thanks @geneticdrift . So I disabled the password to try it out. First call returned nothing, second call returned a lot of entries. 3rd and 4th call returnd nothing, 5th call returned only two entries. 

It is as if the software answered without having finished querying the database.

**Comment by MartyLake** [comment]
Ok I've got something new. I tried a smaller query :
`
echo "search artist \"Oliver\"" | nc mpd_hostname 6600`
got nothing, then wait 2 secs, then retry the query and now I got results ! :O

I could repeat this pattern 3 times in a row.
Waiting 10 seconds yields no result at all.
Spamming as fast as I can the query yields sometimes no results, sometimes results.

**Comment by MaxKellermann** [comment]
What does "returned nothing" mean, exactly? Does MPD close the connection, does it print an empty result, does it omit the "OK" line, do you get an error, ...?

**Comment by geneticdrift** [comment]
@MartyLake 
I tried the following command line and got correct results every time:
<pre>
> for i in 0 1 2 1 0 3 0 ; do sleep $i; echo "search artist \"Pink Floyd\"" | nc -N localhost 6600 | grep Artist | wc -l ; done
43
43
43
43
43
43
43
</pre>
Also consistent when replacing `search artist \"Pink Floyd\"` with `count artist \"Pink Floyd\"`
<pre>
> for i in 0 1 2 1 0 3 0 ; do sleep $i; echo "count artist \"Pink Floyd\"" | nc -N localhost 6600 | grep songs; done
songs: 43
songs: 43
songs: 43
songs: 43
songs: 43
songs: 43
songs: 43
</pre>
Maybe try similar and post the results.

**Comment by mikkoteras** [comment]
I think I'm seeing the same phenomenon on MPD 0.21.15 and 0.21.22 on Windows x64. I also run MPD 0.21.5 on Debian x64, where everything is working properly.

On Windows, the command
search "(artist == 'u2')"
returns an incomplete list of files with varying length but seemingly predictable order.  There should be 167 files in the list, but only 0-30 are returned, varying between consecutive attempts. The zero result is an immediate single-line OK response, not Ack, and the connection stays open.

Basically the same happens when I try
search "((artist == 'u2') AND (album == 'war'))"
This time the result is almost always an empty list, but occasionally 1-2 random files from the album are returned. On Debian and 0.21.5, the result is consistently the full album consisting of ten files.

Interestingly, the "find" command runs just fine.

I'm developing my own, as-yet unpublished client, so I can't point you to it unfortunately, but the effect can easily be seen with a simple telnet connection.

**Comment by MartyLake** [comment]
Thanks for implementing this for win32 as well !!!
